services/src/main/java/io/grpc/services/health/HealthServiceImpl.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.services.health;
+
+import static io.grpc.services.health.HealthOuterClass.HealthCheckResponse;
+import static io.grpc.services.health.HealthOuterClass.HealthCheckResponse.ServingStatus;
+
+import io.grpc.Status;
+import io.grpc.StatusException;
+import io.grpc.stub.StreamObserver;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.annotation.Nullable;
+
+
+final class HealthServiceImpl extends HealthGrpc.AbstractHealth {
+
+  /* Due to the latency of rpc calls, synchronization of the map does not help with consistency.
+   * However, need use ConcurrentHashMap to prevent the possible race condition of currently putting
+   * two keys with a colliding hashCode into the map. */
+  private final Map<String, ServingStatus> statusMap
+      = new ConcurrentHashMap<String, ServingStatus>();
+
+  @Override
+  public void check(HealthOuterClass.HealthCheckRequest request,
+      StreamObserver<HealthCheckResponse> responseObserver) {
+    ServingStatus status = getStatus(request.getService());
+    respond(status, responseObserver);
+  }
+
+  private void respond(ServingStatus status, StreamObserver<HealthCheckResponse> responseObserver) {
```

Comment:
we usually create private methods either for re-using a piece of logic, or breaking down a large method to improve readability. Neither is the case here. It's probably better to merge it with check().